keras/backend/tensorflow_backend.py

```diff
@@ -1033,11 +1033,23 @@ def rnn(step_function, inputs, initial_states,
         new_states: list of tensors, latest states returned by
             the step function, of shape (samples, ...).
     '''
-    ndim = len(inputs.get_shape())
-    assert ndim >= 3, "Input should be at least 3D."
-    axes = [1, 0] + list(range(2, ndim))
-    inputs = tf.transpose(inputs, (axes))
-    input_list = tf.unpack(inputs)
+    if type(inputs) == list:
+        input_list = []
+        for x in inputs:
+            ndim = len(x.get_shape())
+            assert ndim >= 3, "Input should be at least 3D."
+            axes = [1, 0] + list(range(2, ndim))
+            x = tf.transpose(x, (axes))
+            input_list += [tf.unpack(x)]
+        assert len(set(map(len, input_list))) == 1, "All input sequences should be of equal length."
+        input_list = [map(lambda x: x[t], input_list) for t in range(len(input_list[0]))]
```

Comment:
If feel like this should be handled more elegantly in the above loop. This line is not very readable.